### ISSUE TITLE
All the examples in regex::bytes are invalid uses of the API

### ISSUE THREAD

**Issue by jFransham** [rust-lang-nursery/regex#298]
The documentation for `regex::re_set::bytes` and `regex::re_set::unicode` are identical, because they're generated with a macro that is just hardcoded to output the documentation for unicode either way. I'm surprised this isn't caught by doctests (they don't seem to be disabled, although I'm not sure exactly how the configuration for tests in cargo works).

Also, question: why don't the both the `regex::*` and `regex::bytes::*` versions take `AsRef<[u8]>`? Then the documentation would be correct for both. You convert the input to `&[u8]` anyway. I guess it would allow passing invalid UTF8 to `regex::*` but you could at least use those bounds for `regex::bytes::*`.

**Comment by BurntSushi** [comment]
The doc tests don't fail because they all explicitly use `regex::RegexSet`.

But yes, you're otherwise right, the doc examples for `regex::bytes::RegexSet` are incorrect. I wonder if there's a way to make this work while still retaining the macro. Probably not. :-/

> Also, question: why don't the both the regex::* and regex::bytes::* versions take AsRef<[u8]>? Then the documentation would be correct for both. You convert the input to &[u8] anyway. I guess it would allow passing invalid UTF8 to regex::* but you could at least use those bounds for regex::bytes::*.

`regex::Regex` absolutely should not take `AsRef<[u8]>`. While it's true that the `&str` is converted to `&[u8]` internally, the key invariant of `regex::Regex` is that it operates only on UTF-8 and therefore only returns valid UTF-8 indices. More than that, some operations on `regex::Regex` return `&str` slices (and in regex 1.0, `find` will grow that feature as well), and so accepting an `&[u8]` would be quite inappropriate for that case.

I'm not sure why `regex::bytes::Regex` should accept an `&str`. If you have an `&str`, then what's the use case for `regex::bytes::Regex`? Moreover, `regex::bytes::Regex`, like `regex::Regex`, can return `&[u8]` in places. It feels strange and complicated to do that if the caller provided an `&str`.

In short, the two APIs exist precisely because of the distinction between `&str`/`&[u8]`. I think attempting to blur the lines between them is a mistake.

**Comment by jFransham** [comment]
Fair enough, good answer to the question. The documentation point still stands, obviously.
